workers/unity/Packages/io.improbable.worker.sdk/ParameterConversion.cs

```diff
@@ -0,0 +1,140 @@
+                    handle.Free();
+                }
+            }
+
+            public void Dispose()
+            {
+                if (handle.IsAllocated)
+                {
+                    handle.Free();
+                }
+
+                GC.SuppressFinalize(this);
+            }
+
+            public IntPtr Get()
+            {
+                return GCHandle.ToIntPtr(handle);
+            }
+        }
+
+        public static unsafe CEventTrace.SpanId ConvertSpanId(SpanId spanId)
+        {
+            var internalSpanId = new CEventTrace.SpanId();
+            ApiInterop.Memcpy(internalSpanId.Data, spanId.Data, (UIntPtr) 16);
+            return internalSpanId;
+        }
+
+        public static unsafe CEventTrace.Event ConvertEvent(Event eventToConvert)
+        {
+            var internalEvent = new CEventTrace.Event();
+            internalEvent.UnixTimestampMillis = eventToConvert.UnixTimestampMillis;
+            internalEvent.Id = ConvertSpanId(eventToConvert.Id);
+            internalEvent.Data = eventToConvert.Data?.eventData.AddrOfPinnedObject() ?? IntPtr.Zero;
+
+            fixed (byte* eventType = ApiInterop.ToUtf8Cstr(eventToConvert.Type))
+            fixed (byte* eventMessage = ApiInterop.ToUtf8Cstr(eventToConvert.Message))
+            {
+                internalEvent.Type = eventType;
+                internalEvent.Message = eventMessage;
+            }
+
+            return internalEvent;
+        }
+
+        public static unsafe void ConvertEventTracer(EventTracerParameters[] parameters, out CEventTrace.EventTracerParameters[] internalParameters)
+        {
+            List<WrappedGcHandle> handles = new List<WrappedGcHandle>(
+                parameters.Length);
+
+            // ConvertTracerParameters(handles, ...)
+            internalParameters = new CEventTrace.EventTracerParameters[parameters.Length];
+        }
+
+        public static unsafe WrappedGcHandle ConvertTracerParameter(EventTracerParameters parameter, ref CEventTrace.EventTracerParameters internalParameters)
+        {
+            WrappedGcHandle wrappedParameterObject = new WrappedGcHandle(parameter);
+
+            internalParameters.UserData = (void*) wrappedParameterObject.Get();
+            internalParameters.TraceCallback = parameter.TraceCallback == null
+                ? IntPtr.Zero
+                : Marshal.GetFunctionPointerForDelegate(CallbackThunkDelegates.traceCallbackThunkDelegate);
+
+            return wrappedParameterObject;
+        }
+
+        public static void FreeTracerParameterHandles(IEnumerable<WrappedGcHandle> tracerParameterHandles)
+        {
+            foreach (var tracerParameterHandle in tracerParameterHandles)
+            {
+                tracerParameterHandle.Dispose();
+            }
+        }
+
+        private static void ConvertTracerParameters(List<WrappedGcHandle> handles,
+            Dictionary<uint, EventTracerParameters> parameterDict, EventTracerParameters defaultParameters,
```

Comment:
What is the significance of the `uint` as the key?

Reply:
That function was taken from the Worker SDK, I don't think I need the dictionary, they included this since they used the ComponentID to get the relevant object from the dictionary iirc. I think I can change this to just use a List instead 🙂